Settings editors

3nids opened the pull request:
Finally the cool stuff :)

* API
  * settings: Removal of QgsSettingsEntryByValue, only use by reference instead
  * old QgsSettingsTreeWidget has been renamed to QgsSettingsTreeWidgetOld (in app)
  * new classes:
    * QgsSettingsTreeWidget and QgsSettingsTreeModel: tree widget + model to show and edit the settings
    * QgsSettingsEditorWidgetWrapper (pure virtual): a base wrapper class to create and handle a setting editor widget
    * QgsSettingsEditorWidgetWrapperTemplate(pure virtual): the base class for wrapper of settings declared in c++ API (using typed methdos)
    * classes for settings editors: QgsSettingsStringEditorWidgetWrapper, …
    * QgsSettingsEditorWidgetRegistry: a registry of the setting editor wrappers

* UX (in the advanced options):
  * we have the choice of using the old or the new tree until the majority of settings are merged, choice is remembered
  * a "hidden" setting allows to bypass the warning before showing the settings tree
  * settings have dedicated editors (spin box, checkbox, colorbutton, …)
  * the modification of settings is now only achieved when apply is pressed
  * settings not being set yet are in italic, edited settings in red

To come in a next PR (not yet done)
  * [ ] respect constraints of the setting (min/max for int, transparency for colors, …)
  * [ ] tests (will come first)
  * [ ] inform when it's not possible to show and/or edit the setting value (lists and maps for now)
  * [x] ~~filtering~~
  * [ ] reset to defaultValue
  * [ ] delete named node item (e.g. a connection entry)
  * [ ] add missing editor: enums, 64bits integer, list, maps
  * [ ] context menu to delete, reset to original, copy/paste

3nids committed:
simplification of settings API
always pass value by reference (drop of QgsSettingsEntryByValue)
undeprecate setVariantValue

3nids committed:
rename QgsSettingsTreeWidget to QgsSettingsTreeWidget (app)

3nids committed:
add base class and implementation of wrappers for settings editor widgets

3nids committed:
add registry of settings editors wrappers

3nids committed:
add tree view + model to show/edit settings values

3nids committed:
use new settings tree in advanced options
choice can be made to use the old or new tree
also add a hidden setting to bypass warning

3nids committed:
sipify + CMakeLists

3nids committed:
fix dox

3nids commented:
![Screen Recording 2023-04-23 at 07 21 16](https://user-images.githubusercontent.com/127259/233822109-7e339260-ede9-4fc5-8ad9-bbf7df3f6bd9.gif)

3nids committed:
use opactify in colorbutton when possible

3nids committed:
fix dox

nirvn commented on src/gui/settings/qgssettingstreemodel.h:
```diff
@@ -0,0 +1,207 @@
+    void applyChanges();
+
+    /**
+     * Returns settings tree node for given index. Returns root node for invalid index.
```
```suggestion
     * Returns settings tree node for given \a index or the root node if the index is invalid.
```

3nids commented on src/app/options/qgsadvancedoptions.cpp:
```diff
@@ -14,26 +14,41 @@
-    mAdvancedSettingsEditor->show();
     mAdvancedSettingsWarning->hide();
-  } );
+    mGroupBox->layout()->addWidget( createSettingsTreeWidget() );
```
why you reset the setting value here, nothing changed.
`settingShowWarning` is a non GUI exposed setting, so setting it to false will show directly the type of tree you want.

troopa81 commented:
Nice work :+1: 

Just on UI side of things:
- Could we change the color of modified settings? I would say "red" means error for most of users
- I'm wondering if it would'nt be better/possible to display directly the checkbox, instead of true/false and then the checkbox when edit

3nids committed:
Apply suggestions from code review

Co-authored-by: Mathieu Pellerin <nirvn.asia@gmail.com>

3nids committed:
refactor QgsOptionsDialogHighlighWidget to use it as an interface

by removing the inheritance of QObject, an implementation of QWidget can now also inherit from QgsOptionsDialogHighlighWidget and provide the interface to search/highlight texts

3nids committed:
create settings trees on start so texts can be searched

3nids committed:
add proxy model to filter the settings tree

3nids committed:
sipify

3nids committed:
use checkbox for booleans + palette for colors

3nids committed:
better dark theme + fix typo

3nids commented:
Thanks for the feedback @troopa81 

I've implemented the checkbox for the booleans and used the palette colors (see screenshots).
Also, filtering is now done.

<img width="1043" alt="image" src="https://user-images.githubusercontent.com/127259/234972558-4dfe09aa-0db4-43b0-b103-2a0f2b402af7.png">


<img width="1043" alt="image" src="https://user-images.githubusercontent.com/127259/234972075-927ba488-65d1-4972-b187-b2a57911d2c5.png">

3nids committed:
add missing Q_OBJECT macro

3nids committed:
remove useless method

3nids committed:
use public method

troopa81 commented:
Thanks @3nids , it looks better :+1:
